airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/DebeziumRecordIterator.java

```diff
@@ -0,0 +1,144 @@
+ * started. Because the publisher might publish more records between the consumer sending this
+ * signal and the publisher acutally shutting down, the consumer must stay alive as long as the
+ * publisher is not closed or if there are any new records for it to process (even if the publisher
+ * is closed).
+ */
+public class DebeziumRecordIterator extends AbstractIterator<ChangeEvent<String, String>>
+    implements AutoCloseableIterator<ChangeEvent<String, String>> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DebeziumRecordIterator.class);
+
+  private static final TimeUnit SLEEP_TIME_UNIT = TimeUnit.SECONDS;
+  private static final int SLEEP_TIME_AMOUNT = 5;
+
+  private final LinkedBlockingQueue<ChangeEvent<String, String>> queue;
+  private final Optional<TargetFilePosition> targetFilePosition;
+  private final Supplier<Boolean> publisherStatusSupplier;
+  private final VoidCallable requestClose;
+
+  public DebeziumRecordIterator(LinkedBlockingQueue<ChangeEvent<String, String>> queue,
+                                Optional<TargetFilePosition> targetFilePosition,
+                                Supplier<Boolean> publisherStatusSupplier,
+                                VoidCallable requestClose) {
+    this.queue = queue;
+    this.targetFilePosition = targetFilePosition;
+    this.publisherStatusSupplier = publisherStatusSupplier;
+    this.requestClose = requestClose;
+  }
+
+  @Override
+  protected ChangeEvent<String, String> computeNext() {
+    // keep trying until the publisher is closed or until the queue is empty. the latter case is
+    // possible when the publisher has shutdown but the consumer has not yet processed all messages it
+    // emitted.
+    while (!MoreBooleans.isTruthy(publisherStatusSupplier.get()) || !queue.isEmpty()) {
+      final ChangeEvent<String, String> next;
+      try {
+        next = queue.poll(SLEEP_TIME_AMOUNT, SLEEP_TIME_UNIT);
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+
+      // if within the timeout, the consumer could not get a record, it is time to tell the producer to
+      // shutdown.
+      if (next == null) {
+        requestClose();
+        LOGGER.info("no record found. polling again.");
+        continue;
+      }
+
+      // if the last record matches the target file position, it is time to tell the producer to shutdown.
+      if (shouldSignalClose(next)) {
+        requestClose();
+      }
+
+      return next;
+    }
+    return endOfData();
+  }
+
+  @Override
+  public void close() throws Exception {
+    requestClose.call();
+  }
+
+  private boolean shouldSignalClose(ChangeEvent<String, String> event) {
+    if (targetFilePosition.isPresent()) {
+      String file = Jsons.deserialize(event.value()).get("source").get("file").asText();
+      int position = Jsons.deserialize(event.value()).get("source").get("pos").asInt();
+      if (file.equals(targetFilePosition.get().fileName)) {
```

Comment:
writing the above example made me think about this line.

are we comparing file names here? (slightly confused since we are compared `file` to `filename`) what happens if the names don't match up? should we error?

Reply:
There can be multiple files, we target the file with the latest records as end point. If the file name doesn't match then we are behind the latest file